README.md

```diff
@@ -278,3 +278,50 @@ jsonPayload.verb="stacktraceanalyzer"
 3. `stack_trace_interval_seconds` signifies the duration in seconds between each stack trace collection event. Setting `stack_trace_interval_seconds: 600` will collect the stack traces every 600 seconds (10 minutes).
 
 Here is the related PyPI package: https://pypi.org/project/cloud-tpu-diagnostics.
+
+## Ahead of Time Compilation (AOT)
+To compile your training run ahead of time, we provide a tool `train_compile.py`. This tool allows you to compile the main `train_step` in `train.py` for target hardware (e.g. a large number of v5e devices) without using the target hardware, and instead you may use only a CPU or a single VM from a different family. This compilation helps with two main goals:
+
+* It will flag any out of memory (OOM) information, such as when the `per_device_batch_size` is set too high, with an identical OOM stack trace as if it was compiled on the target hardware.
+
+* The ahead of time compilation can be saved and then loaded for fast startup and restart times on the target hardware.
+
+The tool `train_compile.py` is tightly linked to `train.py` and uses the same configuration file `configs/base.yml`. Although you don't need to run on a TPU, you do need to install `jax[tpu]` in addition to other dependenices, so we recommend running `setup.sh` to install these if you have not already done so. 
+
+### Example AOT 1: Compile ahead of time basics
+After installing the dependencies listed above, you are ready to compile ahead of time:
+```
+# Run the below on a single machine, e.g. a CPU
+python3 MaxText/train_compile.py MaxText/configs/base.yml compile_topology=v5e-256 compile_topology_num_slices=2 \ 
+global_parameter_scale=16 per_device_batch_size=4
+```
+
+This will compile a 16B parameter MaxText model on 2 v5e pods. 
+
+### Example AOT 2: Save compiled function, then load and run it
+Here is an example that saves then loads the compiled `train_step`, starting with the save:
+
+**Step 1: Run AOT and save compiled function**
+```
+# Run the below on a single machine, e.g. a CPU
+export LIBTPU_INIT_ARGS="--xla_enable_async_all_gather=true" 
+python3 MaxText/train_compile.py MaxText/configs/base.yml compile_topology=v5e-256 \
+compile_topology_num_slices=2 \
+compiled_trainstep_file=my_compiled_train.pickle global_parameter_scale=16 \
+per_device_batch_size=4 steps=10000 learning_rate=1e-3
+```
+
+**Step 2: Run train.py and load the compiled function**
+
+To load the compiled train_step, you just need to pass `compiled_trainstep_file=my_compiled_train.pickle` into `train.py`:
+```
+# Run the below on each host of the target hardware, e.g. each host on 2 slices of v5e-256
+python3 MaxText/train.py MaxText/configs/base.yml run_name=example_load_compile \
```

Comment:
shouldn't we have `export LIBTPU_INIT_ARGS="--xla_enable_async_all_gather=true"` here?

Reply:
(Don't we need a learning_rate here?)

Reply:
Done.